homeassistant/helpers/config_validation.py

```diff
@@ -402,3 +402,9 @@ def validator(value):
     [vol.Any(SERVICE_SCHEMA, _SCRIPT_DELAY_SCHEMA, EVENT_SCHEMA,
              CONDITION_SCHEMA)],
 )
+
+
+def string_schema(*args: Sequence) -> Dict:
+    """Return a schema dict of Required values or Optional values."""
+    return {val if isinstance(val, vol.Optional) else vol.Required(val): string
+            for val in args}
```

Comment:
Don't be changing my PRs with your shiz! :-P

On Fri, Sep 2, 2016 at 3:37 PM, Johann Kellerman <notifications@github.com>
wrote:

> In homeassistant/helpers/config_validation.py
> <https://github.com/home-assistant/home-assistant/pull/3123#discussion_r77415213>
> :
>
> > @@ -402,3 +402,9 @@ def validator(value):
> >      [vol.Any(SERVICE_SCHEMA, _SCRIPT_DELAY_SCHEMA, EVENT_SCHEMA,
> >               CONDITION_SCHEMA)],
> >  )
> > +
> > +
> > +def string_schema(*args: Sequence) -> Dict:
> > +    """Return a schema dict of Required values or Optional values."""
> > +    return {val if isinstance(val, vol.Optional) else vol.Required(val): string
> > +            for val in args}
>
> It can cut down you ODB tracker schema like this:
>
> PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
>     vol.Required(CONF_CLIENT_ID): cv.string,
>     vol.Required(CONF_SECRET): cv.string,
>     vol.Required(CONF_USERNAME): cv.string,
>     vol.Required(CONF_PASSWORD): cv.string,
>     vol.Optional(CONF_DEVICES): vol.All(cv.ensure_list, [cv.string])
> })
>
> to
>
> PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
>     string_schema(CONF_CLIENT_ID, CONF_SECRET, CONF_USERNAME, CONF_PASSWORD)
> ).extend({
>     vol.Optional(CONF_DEVICES): vol.All(cv.ensure_list, [cv.string])
> })
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/3123/files/9d441d3b8993b78c49f4a9b22ac2a5255700f75a#r77415213>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC2fZSKX41gWgZHomumxFCNhwIPFLJ67ks5qmJcGgaJpZM4JzJse>
> .
>